lib/os/onoff.c

```diff
@@ -154,320 +428,175 @@ int onoff_request(struct onoff_manager *mgr,
-	 */
-	notify_one(mgr, releaser, res);
-	if (notify_clients) {
-		notify_all(mgr, &clients, client_res);
-	} else if (start) {
-		mgr->transitions->start(mgr, onoff_start_notify);
-	}
+	return rv;
 }
 
-int onoff_release(struct onoff_manager *mgr,
-		  struct onoff_client *cli)
+int onoff_reset(struct onoff_manager *mgr,
+		struct onoff_client *cli)
 {
-	bool stop = false;      /* invoke stop transition */
-	bool notify = false;    /* do client notification */
+	bool reset = false;
 	int rv = validate_args(mgr, cli);
 
+	if ((rv >= 0)
+	    && (mgr->transitions->reset == NULL)) {
+		rv = -ENOTSUP;
+	}
+
 	if (rv < 0) {
 		return rv;
 	}
 
 	k_spinlock_key_t key = k_spin_lock(&mgr->lock);
+	u32_t state = mgr->flags & ONOFF_STATE_MASK;
 
-	if ((mgr->flags & ONOFF_HAS_ERROR) != 0) {
-		rv = -EIO;
-		goto out;
-	}
-
-	u32_t state = mgr->flags & ST_MASK;
-
-	switch (state) {
-	case ST_ON:
-		/* Stay on if release leaves a client. */
-		if (mgr->refs > 1U) {
-			notify = true;
-			rv = 1;
-			break;
-		}
-
-		/* Reject if in non-thread context but stop could
-		 * wait
-		 */
-		if ((k_is_in_isr() || k_is_pre_kernel())
-		    && ((mgr->flags & ONOFF_STOP_SLEEPS) != 0)) {
-			rv = -EWOULDBLOCK;
-			break;
-		}
-
-		stop = true;
-
-		set_service_state(mgr, ST_TO_OFF);
-		mgr->releaser = cli;
-		rv = 2;
+	rv = state;
 
-		break;
-	case ST_TO_ON:
-		rv = -EBUSY;
-		break;
-	case ST_OFF:
-	case ST_TO_OFF:
+	if ((state & ONOFF_FLAG_ERROR) == 0) {
 		rv = -EALREADY;
-		break;
-	default:
-		rv = -EINVAL;
-	}
-
-out:
-	if (notify) {
-		mgr->refs -= 1U;
+	} else {
+		reset = (state != ONOFF_STATE_RESETTING);
+		sys_slist_append(&mgr->clients, &cli->node);
 	}
 
-	k_spin_unlock(&mgr->lock, key);
-
-	if (stop) {
-		__ASSERT_NO_MSG(mgr->transitions->stop != NULL);
-		mgr->transitions->stop(mgr, onoff_stop_notify);
-	} else if (notify) {
-		notify_one(mgr, cli, 0);
+	if (reset) {
+		process_event(mgr, EVT_RECHECK, key);
+	} else {
+		k_spin_unlock(&mgr->lock, key);
 	}
 
 	return rv;
 }
 
-static void onoff_reset_notify(struct onoff_manager *mgr,
-			       int res)
+int onoff_cancel(struct onoff_manager *mgr,
+		 struct onoff_client *cli)
 {
+	if ((mgr == NULL) || (cli == NULL)) {
+		return -EINVAL;
+	}
+
+	int rv = -EALREADY;
 	k_spinlock_key_t key = k_spin_lock(&mgr->lock);
-	sys_slist_t clients = mgr->clients;
+	u32_t state = mgr->flags & ONOFF_STATE_MASK;
 
-	/* If the reset failed clear the transition flag but otherwise
-	 * leave the state unchanged.
-	 *
-	 * If it was successful clear the reference count and all
-	 * flags except capability flags (sets to ST_OFF).
-	 */
-	if (res < 0) {
-		mgr->flags &= ~ST_TRANSITION;
-	} else {
-		__ASSERT_NO_MSG(mgr->refs == 0U);
-		mgr->refs = 0U;
-		mgr->flags &= SERVICE_CONFIG_FLAGS;
+	if (sys_slist_find_and_remove(&mgr->clients, &cli->node)) {
+		__ASSERT_NO_MSG((state == ONOFF_STATE_TO_ON)
+				|| (state == ONOFF_STATE_TO_OFF)
+				|| (state == ONOFF_STATE_RESETTING));
+		rv = state;
```

Comment:
what about cleaning client state here. Otherwise `sys_notify_fetch_result` will return -EAGAIN.

Reply:
Yes, it will.  Why is that wrong?  In an earlier review you asked that cancel not record a completion because the natural value for it `-ECANCELED` might conflict with a legitimate result.  So the client data is left in an imcomplete state, for which no result is available.

Reply:
ok, lets leave it. I was thinking about a way to determine who is currently owning the client and fetch result could be used if state was cleaned here and in cancel_or_release.